protocols/mempool-overview.md

```diff
@@ -134,10 +139,134 @@ Committed transactions are not in the mempool:
 
 In blockchain, a tx is (or not) valid in a given state.
 That is a transaction can be valid (or not) at a given height of the ledger.
-To model this, consider that $p.ledger.valid(tx)$ is such a check for the current height of the ledger at process $p$ (ACBI call).
+To model this, consider that $p.ledger.valid(tx)$ is such a check for the current height of the ledger at process $p$ (ABCI call).
 Our third invariant is that only valid transactions are present in the mempool:  
 **INV3.** $\forall tx, \forall p \in Correct. \square(tx \in p.mempool \implies p.ledger.valid(tx))$
 
 Finally, we require some progress from the mempool.
 Namely, if a transaction appears at a correct process then eventually it is committed or forever invalid.  
 **INV4** $\forall tx. \forall p \in Correct. \square(tx \in p.mempool \implies \lozenge\square(tx \in p.committed \vee \neg p.ledger.valid(tx)))$
+
+The above invariant ensures that if a transaction enters the mempool, then it eventually leaves it at all the correct processes.
+It requires from the application that the validity of a transaction converges toward some value.
+That is, in every run, there is a height after which $valid(tx)$ always returns the same value.
+This happens in [Ethereum](https://github.com/ethereum/go-ethereum/blob/5c51ef8527c47268628fe9be61522816a7f1b395/light/txpool.go#L401) as a transaction is always valid until a transaction from the same address executes with the same or higher nonce.
+A simpler way to satisfy this requirement is by having $valid(tx)$ deterministic and stateless (e.g., a syntactic check).
+
+**Practical considerations.**
+In practice, as it requires to traverse the whole ledger, INV2 is too expensive.
+Instead, we would like to maintain this only over the last $\alpha$ committed transactions, for some parameter $\alpha$.
+Given a process $p$, we write $p.lcommitted$ the last $\alpha$ committed transactions at $p$.
+Invariant INV2 is replaced with:  
+**INV2a.** $\forall tx. \forall p \in Correct. \square(tx \in p.lcommitted \implies tx \notin p.mempool)$
+
+Another practical concern is with INV3.
+This invariant requires to have a green light from the client application before adding a transaction to the mempool.
+For efficiency, such a validation needs to be made at most $\beta$ times per transaction at each height, for some parameter $\beta$.
+Ideally, $\beta$ equals $1$.
+In practice, $\beta = f(T)$ for some function f of the maximal number of transactions T submitted between two heights.
+Given some transaction $tx$, variable $p.valid[tx]$ tracks the number of times the application was asked at the current height.
```

Comment:
I think $p.valid$ here is not a great name. I would use something less "definitive", like $p.checks$ or something like that.

Reply:
All in all, this is almost an implementation detail. Not sure whether we need this modeled here.